Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -523,6 +523,90 @@ A similar approach works. Answer the following questions:
 - Did you need to write custom code to disable the defenses? How much time did you need?
 - What is your assessment of the difficulty of bypassing the mechanisms??
 
+### Testing Reverse Engineering Tools Detection (MSTG-RESILIENCE-4)
+
+#### Overview
+
+The presence of reverse engineering tools on the device may indicate that the user is attempting to reverse engineer the app. Some reverse engineering tools can only run on a jailbroken device, force the app into debugging mode or start a service on the mobile phone. Different ways would need to be implemented in the mobile app to detect a reverse engineering attack and react to it, like terminating the app.
+
+#### Detection Methods
+
+You can detect popular reverse engineering tools that have been installed in an unmodified form by looking for associated application packages, files, processes, or other tool-specific modifications and artifacts. In the following examples, we'll discuss different ways to detect the Frida instrumentation framework, which is used extensively in this guide and also in the real world. Other tools, such as Cydia Substrate / Cycript, can be detected similarly. Note that injection/hooking/DBI (Dynamic Binary Instrumentation) tools can often be detected implicitly, through run time integrity checks, which are discussed below.
+
+For instance, in its default configuration on a jailbroken device, Frida runs as frida-server. When you explicitly attach to a target app (e.g. via frida-trace or the Frida REPL), Frida injects a frida-agent into the memory of the app. Therefore, you may expect to find it there after attaching to the app (and not before). On Android you can verify this, by grepping for the string "frida" in the memory (`maps`) of the process ID in the `proc` directory.
+On iOS the `proc` directory is not available, but you can list the loaded dynamic libraries in an app with the function `_dyld_image_count`.
```

Comment:
- Added "however" to link to the prev. paragraph.
- Just wondering if `_dyld_image_count` is a public or private API? Does the Apple allow this on Store apps?

```suggestion
However, on iOS the `proc` directory is not available, but you can list the loaded dynamic libraries in an app with the function `_dyld_image_count`.
```

Reply:
_dyld_image_count is used in the iOS Security Suite. And this library was already implemented in at least one public app that passed the app review process.